MapboxCoreNavigation/Router.swift

```diff
@@ -93,12 +102,68 @@ protocol InternalRouter: class {
     
     var isRerouting: Bool { get set }
     
+    var isRefreshing: Bool { get set }
+    
     var directions: Directions { get }
     
     var routeProgress: RouteProgress { get set }
 }
 
 extension InternalRouter where Self: Router {
+    
+    func refreshAndCheckForFasterRoute(from location: CLLocation, routeProgress: RouteProgress) {
+        if refreshesRoute {
+            refreshRoute(from: location, legIndex: routeProgress.legIndex) {
+                self.checkForFasterRoute(from: location, routeProgress: routeProgress)
+            }
+        } else {
+            checkForFasterRoute(from: location, routeProgress: routeProgress)
+        }
+    }
+    
+    func refreshRoute(from location: CLLocation, legIndex: Int, completion: @escaping ()->()) {
+        guard refreshesRoute else {
+            completion()
+            return
+        }
+        
+        guard let lastRouteRefresh = lastRouteRefresh else {
+            self.lastRouteRefresh = location.timestamp
+            completion()
+            return
+        }
+        
+        guard location.timestamp.timeIntervalSince(lastRouteRefresh) >= RouteControllerProactiveReroutingInterval else {
+            completion()
+            return
+        }
+        
+        if isRefreshing {
+            completion()
+            return
+        }
+        isRefreshing = true
+        
+        directions.refresh(route: route,
+                           currentLegIndex: legIndex,
+                           completionHandler: { [weak self] (session, result) in
+                            defer {
+                                self?.isRefreshing = false
+                                self?.lastRouteRefresh = nil
+                                completion()
+                            }
+                            
+                            guard case let .success(response) = result else {
+                                return
+                            }
+                            
+                            guard let route = response.route else {
```

Comment:
As of mapbox/mapbox-directions-swift@3941c256c9938b56ffaea5a8b7443a0d1c24b0e7, `RouteRefreshResponse.route` is merely a skeletal RefreshedRoute, not a fully merged Route. We need to call `Route.refreshLegAttributes(from:)` here.

Reply:
#2591 will take care of this feedback.